wip: create a flat options object to pass to deps

isaacs opened the pull request:
Related to [npm/rfcs#102](https://github.com/npm/rfcs/pull/102).

wip, needs tests.

Commit:
npm-pick-manifest@5.0.0

Commit:
npm-registry-fetch@6.0.0

Commit:
test: removing test for no scoped deps, dropping constraint for v7 of cli

claudiahdz committed:
ssri@7.1.0

claudiahdz committed:
npm-package-arg@8.0.0

claudiahdz committed:
which@2.0.2

claudiahdz committed:
cacache@14.0.0

claudiahdz committed:
tar@6.0.1

Commit:
pacote@10.3.1

Commit:
feat: convert figgy config into POJO, when passed to pacote

isaacs committed:
wip: create a flat options object to pass to deps

evocateur commented on lib/config/flat-options.js:
```diff
@@ -0,0 +1,216 @@
+
+const flatOptions = npm => ({
+  // Note that many of these do not come from configs or cli flags
+  // per se, though they may be implied or defined by them.
```
Speaking as someone who maintains a CLI (lerna) that implements npm publishing using current `libnpm*` packages (with an alarming amount of copypasta from npm's source) _and_ attempts to stay as consistent as possible with npm CLI patterns, extracting "How npm reads and serializes its configuration" into a separate package would be a great boon.

I realize it's extremely low on the priority list, but I just wanted to register the sentiment while it's top-of-mind, and before I inevitably feel the urge to make more copypasta...

isaacs replied:
Yeah, that would be really nice, I agree.  (Also speaking as someone who maintains and uses a lot of npm deps ;)

We had once upon a time a package called `npmconf` that housed all of the config reading and management as  well as all the default values.  The problem was that every time we wanted to add or modify a config value, it meant an update in a bunch of disparate places, and it was kind of a pain to manage, so we folded it back in.  This gets more convoluted when we have a bunch of other modules expecting to get a `preferOnline` option and need to know to convert that from `config.get('prefer-online')` (or even `config.get('cache-min')<=0`).  The goal of figgy-pudding was to make this easier, but in practice, has made it much harder, which is why we're moving to the simplest possible approach, and from there can build up something a bit more maintainable.  Just taking an inventory of all our configs as part of this upgrade process has been really enlightening.  The clever technical solution didn't actually get at the root of the problem, and doing the work to dredge it all up and clean the actual mess has been sorely needed.

Some day, I'd really like to consider dropping nopt in favor of something a bit more simple and declarative, like jackspeak or even yargs, and then have a package that _doesn't_ handle defaults, but _just_ reads and updates config files, so that the set of defaults and expected types can still live in npm/cli maybe.  It's a weird sort of animal to cut in half, though, because I do want to be able to provide some feedback to the user that `--umask=notanumber` is an invalid config value.  So I'm not sure exactly what that'll look like.  We may be in this "create a pojo and share it" mode for a while, most likely at least through the v7 release line.  It just makes it so much easier to manage our deps if they can get an options object like any other JavaScript module out there does, without the extra translation layer of fp.

isaacs commented:
Suggestion was made (and I think it's a good ideal) to `Object.freeze()` this object.  https://github.com/npm/rfcs/pull/102#discussion_r379230392

isaacs committed:
update several deps to non-figgy-pudding versions

- cacache@15.0.0
- npm-pick-manifest@6.0.0
- npm-registry-fetch@7.0.0
- pacote@11.0.0

isaacs commented:
The good bits of this landed already in the feature/update-deps-v1 branch, so I'm closing this PR.
